test/e2e/clusterctl_upgrade.go

```diff
@@ -71,6 +72,9 @@ type ClusterctlUpgradeSpecInput struct {
 	// InitWithKubernetesVersion can be used to override the INIT_WITH_KUBERNETES_VERSION e2e config variable with a specific
 	// Kubernetes version to use to create the secondary management cluster, e.g. `v1.25.0`
 	InitWithKubernetesVersion string
+	// InitWithWorkloadKubernetesVersion can be used to override the INIT_WITH_WORKLOAD_KUBERNETES_VERSION e2e config variable with a specific
+	// Kubernetes version to use to create the workload cluster, e.g. `v1.25.0`
+	InitWithWorkloadKubernetesVersion string
```

Comment:
The idea behind the init suffix is that those are fields related to the clusterctl init call when we install the old provider version (init vs that we upgrade them later).

I think this doesn't work well with the Kubenetes version of the workload cluster that is created with those old provider and later reconciled by the new providers. That's why I suggested WorkloadKubernetesVersion to align to WorkloadFlavor which is what we use to specify the flavor of this same cluster.

Reply:
So for the field I would definitely prefer WorkloadKubernetesVersion. The env variables are a bit of problem as the existing ones are not scoped to this test at all.

I would suggest to expose it via env var as "CLUSTERCTL_UPGRADE_WORKLOAD_KUBERNETES_VERSION". This can be also cherry-picked to the v1.4 branch the same way. I think then it would be good to have a follow-up PR to cleanup the other env vars and also add it to the migration guide (today those env vars are not used by core CAPI, just by other infra providers)

Reply:
I will open a separate issue to track it later today (will come back to here when I do so and refer to it)

Reply:
Sounds good. Feel free to resolve then

Reply:
Coming back to this again, issue to track: https://github.com/kubernetes-sigs/cluster-api/issues/8535